File: src/console/command/impl/server/ServerCommand.php
<?php

namespace pocketcloud\cloud\console\command\impl\server;

use pocketcloud\cloud\console\command\parameter\def\BoolParameter;
use pocketcloud\cloud\console\command\parameter\def\IntegerParameter;
use pocketcloud\cloud\console\command\parameter\def\MultipleTypesParameter;
use pocketcloud\cloud\console\command\parameter\def\ServerParameter;
use pocketcloud\cloud\console\command\parameter\def\ServerGroupParameter;
use pocketcloud\cloud\console\command\parameter\def\StringParameter;
use pocketcloud\cloud\console\command\parameter\def\StringEnumParameter;
use pocketcloud\cloud\console\command\parameter\def\TemplateParameter;
use pocketcloud\cloud\console\command\Command;
use pocketcloud\cloud\console\command\sender\ICommandSender;
use pocketcloud\cloud\console\command\SubCommand;
use pocketcloud\cloud\console\screen\impl\ServerConsoleMonitorScreen;
use pocketcloud\cloud\console\screen\ScreenManager;
use pocketcloud\cloud\network\packet\data\ServerCommandExecutionResult;
use pocketcloud\cloud\server\CloudServer;
use pocketcloud\cloud\server\CloudServerManager;
use pocketcloud\cloud\server\util\ServerStatus;
use pocketcloud\cloud\util\FileUtils;
use pocketcloud\cloud\util\FormatUtils;

final class ServerCommand extends Command {

    public function __construct() {
        parent::__construct("server", "Manage the cloud's servers", ["srv"]);

        $this->registerSubCommand(SubCommand::fromClosure("start", $this->handleStartSub(...), ["start"])
            ->addParameter(new TemplateParameter("template", false))
            ->addParameter(new IntegerParameter("amount", true, function (int $number): int {
                if ($number < 0) return 1;
                return min(50, $number);
            })));

        $this->registerSubCommand(SubCommand::fromClosure("stop", $this->handleStopSub(...), ["stop"])
            ->addParameter(new MultipleTypesParameter("server", [
                new ServerParameter("server", false),
                new TemplateParameter("template", false),
                new ServerGroupParameter("group", false),
                new StringEnumParameter("all", ["all"], false, false)
            ], false))
            ->addParameter(new BoolParameter("forcefully", true)));

        $this->registerSubCommand(SubCommand::fromClosure("send", $this->handleSendSub(...), ["send", "execute"])
            ->addParameter(new ServerParameter("server", false))
            ->addParameter(new StringParameter("commandLine", false, true)));

        $this->registerSubCommand(SubCommand::fromClosure("list", $this->handleListSub(...))
            ->addParameter(new MultipleTypesParameter("template", [
                new TemplateParameter("template", true),
                new ServerGroupParameter("group", true)
            ], true)));

        $this->registerSubCommand(SubCommand::fromClosure("info", $this->handleInfoSub(...))
            ->addParameter(new ServerParameter("server", false)));

        $this->registerSubCommand(SubCommand::fromClosure("save", $this->handleSaveSub(...), ["save"])
            ->addParameter(new ServerParameter("server", false)));

        $this->registerSubCommand(SubCommand::fromClosure("screen", $this->handleScreenSub(...), ["screen"])
            ->addParameter(new ServerParameter("server", false)));
    }

    public function run(ICommandSender $sender, string $label, array $args, ?SubCommand $subCommand = null): bool {
        return true;
    }

    private function handleStartSub(ICommandSender $sender, string $label, array $args): bool {
        $template = $args["template"];
        $amount = $args["amount"] ?? 1;
        CloudServerManager::getInstance()->start($template, $amount);
        return true;
    }

    private function handleStopSub(ICommandSender $sender, string $label, array $args): bool {
        $server = $args["server"];
        $forcefully = $args["forcefully"] ?? false;

        if (is_string($server) && $server == "all") CloudServerManager::getInstance()->stopAll($forcefully);
        else CloudServerManager::getInstance()->stop($server, $forcefully);
        return true;
    }

    private function handleSendSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var CloudServer $server */
        $server = $args["server"];
        $commandLine = $args["commandLine"];
        $server->executeCommand($commandLine)->then(function (ServerCommandExecutionResult $result) use($sender, $server): void {
            $sender->success("Successfully ran the command on §b{}§r, server responded with the following messages:", $server->getName());
            if (empty($result->getMessages())) $sender->success("§cNone");
            foreach ($result->getMessages() as $message) {
                foreach (explode("\n", $message) as $msgPart) {
                    $sender->success(trim($msgPart));
                }
            }
        })->failure(fn() => $sender->warn("The command execution request ran out. §8(§b{}§8)", $commandLine));
        return true;
    }

    private function handleListSub(ICommandSender $sender, string $label, array $args): bool {
        $template = $args["template"] ?? null;
        if (empty(CloudServerManager::getInstance()->getAll($template))) $sender->info("§cNo servers running.");
        $sender->info("Servers §8(§b{}§8/§b{}§8)§r:", count($servers = CloudServerManager::getInstance()->getAll($template)), $template?->getName() ?? "All");
        foreach ($servers as $server) {
            $sender->info(FormatUtils::implodeWithKeys(
                $server->write(),
                " §8| §r",
                "§8: §b",
                fn(string $key) => ucfirst($key),
                function (string $key, mixed $value) use ($server): mixed {
                    if ($key == "serverStatus") {
                        return ServerStatus::fromName($value)?->getDisplay() ?? $value;
                    } else if ($key == "template") {
                        return $server->getTemplateName() . ($server->getTemplate()?->getParentServerGroup() !== null ? " §8(§e" . $server->getTemplate()->getParentServerGroup()->getName() . "§8)" : "");
                    }

                    return $value;
                },
                "uuid", "id", "internalStorage", "tps", "avgTps", "memoryUsage", "memoryPeak", "memoryLimit", "cpuUsage"
            ));
        }

        return true;
    }

    private function handleInfoSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var CloudServer $server */
        $server = $args["server"];
        $formatted = FormatUtils::implodeWithKeys(
            array_merge($server->write(), ["path" => $server->getPath(), "channel" => $server->getServerClient()?->getAddress()]),
            "\n",
            "§8: §b",
            function (string $key): string {
                $potentialNewKey = preg_replace("/(?<!^)([A-Z])/", " $1", ucfirst($key));
                return match ($key) {
                    "serverStatus" => "Status",
                    "avgTps" => "Average TPS",
                    default => $potentialNewKey
                };
            },
            function (string $key, mixed $value) use($server): mixed {
                if (is_array($value)) return FileUtils::encodeJson($value, JSON_INVALID_UTF8_IGNORE | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);

                if ($key == "name") {
                    return "§b" . $value . " §8(§b" . $server->getId() . "§8/§b" . $server->getTemplateName() . "§8)";
                } else if ($key == "tps" || $key == "avgTps") {
                    return FormatUtils::tps($value);
                } else if (str_contains($key, "memory")) {
                    return FormatUtils::bytes($value);
                } else if (str_contains($key, "Usage")) {
                    return FormatUtils::usagePercentage($value);
                } else if ($key == "serverStatus") {
                    return $server->getServerStatus()?->getDisplay() ?? "Not Applied";
                }

                return $value;
            },
            "template", "id"
        );

        foreach (explode("\n", $formatted) as $line) {
            $sender->info($line);
        }

        return true;
    }

    private function handleSaveSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var CloudServer $server */
        $server = $args["server"];
        $sender->info("Saving §b{}§r...", $server->getName());
        $start = microtime(true) * 1000;
        CloudServerManager::getInstance()->save($server)->then(fn() => $sender->success("Successfully §asaved §b{}§r. §8(§rTook §b{}ms§8)", $server->getName(), round((microtime(true) * 1000) - $start, 3)))
            ->failure(fn(?string $reason) => $sender->warn("Failed to save §b{}§r: §c{}", $server->getName(), $reason ?? "No reason applied"));
        return true;
    }

    private function handleScreenSub(ICommandSender $sender, string $label, array $args): bool {
        /** @var CloudServer $server */
        $server = $args["server"];
        ScreenManager::getInstance()->setCurrentScreen(new ServerConsoleMonitorScreen($server->getName()));
        return true;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/impl/server/ServerCommand.java
package de.pocketcloud.cloud.console.command.impl.server;

import de.pocketcloud.api.component.server.ICloudServer;
import de.pocketcloud.api.search.ServerSearchQuery;
import de.pocketcloud.api.template.TemplateType;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.command.Command;
import de.pocketcloud.cloud.console.command.ctx.CommandContext;
import de.pocketcloud.cloud.console.command.desc.CommandDescription;
import de.pocketcloud.cloud.console.command.flag.CommandFlag;
import de.pocketcloud.cloud.console.command.parameter.def.*;
import de.pocketcloud.cloud.console.command.sender.CommandSender;
import de.pocketcloud.cloud.console.command.sub.SubCommand;
import de.pocketcloud.cloud.console.screen.impl.ServerConsoleMonitorScreen;
import de.pocketcloud.cloud.server.CloudServer;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.common.util.FormatUtils;
import de.pocketcloud.common.util.NumberUtils;
import de.pocketcloud.shared.component.software.ServerSoftware;

import java.time.Duration;
import java.time.Instant;
import java.util.Collection;
import java.util.Comparator;

@CommandDescription(name = "server", description = "Manage the cloud servers", aliases = {"srv", "service"})
public final class ServerCommand extends Command {

    @Override
    public void prepare() {
        registerSubCommand(SubCommand.lambda("start", this::handleStartSub, true, subCommand -> {
            subCommand.addParameter(new TemplateParameter("template", false))
                    .addParameter(new IntegerParameter("amount", true, i -> Math.clamp(i, 1, 100)));
        }));

        registerSubCommand(SubCommand.lambda("stop", this::handleStopSub, true, subCommand -> {
            subCommand.addParameter(new MultipleTypesParameter("server", false, new TemplateParameter("template", true),
                            new ServerParameter("server", true),
                            new TemplateTypeParameter("server", true),
                            new ServerGroupParameter("server_group", true),
                            new TemplateTypeParameter("template_type", true),
                            new SoftwareParameter("software", true)))
                    .addFlag(CommandFlag.shortFlag("f"));
        }));

        registerSubCommand(SubCommand.lambda("stopall", this::handleStopAllSub, true, subCommand -> {
            subCommand.addFlag(CommandFlag.shortFlag("f"))
                    .addFlag(CommandFlag.shortFlag("y"));
        }));

        registerSubCommand(SubCommand.lambda("dispatch", this::handleDispatchSub, true, subCommand -> {
            subCommand.addParameter(new ServerParameter("server", false))
                    .addParameter(new StringParameter("command", false, true));
        }));

        registerSubCommand(SubCommand.lambda("save", this::handleSaveSub, true, subCommand -> {
            subCommand.addParameter(new ServerParameter("server", false));
        }));

        registerSubCommand(SubCommand.lambda("info", this::handleInfoSub, true, subCommand -> {
            subCommand.addParameter(new ServerParameter("server", false));
        }));

        registerSubCommand(SubCommand.lambda("screen", this::handleScreenSub, true, subCommand -> {
            subCommand.addParameter(new ServerParameter("server", false));
        }));

        registerSubCommand(SubCommand.lambda("list", this::handleListSub, true, subCommand -> {
            subCommand.addParameter(new MultipleTypesParameter("filter", true, new TemplateParameter("template", false),
                    new TemplateTypeParameter("template_type", true),
                    new ServerGroupParameter("server_group", true),
                    new SoftwareParameter("software", true)));
        }));
    }

    @Override
    public boolean run(CommandSender sender, CommandContext ctx) {
        return true;
    }

    public boolean handleStartSub(CommandSender sender, CommandContext ctx) {
        int amount = ctx.hasArg("amount") ? ctx.argInt("amount") : 1;
        PocketCloud.instance().servers().start(ctx.arg("template", Template.class), amount);
        return true;
    }

    public boolean handleStopSub(CommandSender sender, CommandContext ctx) {
        Object obj = ctx.arg("server");
        boolean forcefully = ctx.hasFlag("f");
        if (obj instanceof CloudServer server) {
            PocketCloud.instance().servers().stop(server, forcefully);
        } else if (obj instanceof Template template) {
            PocketCloud.instance().servers().stop(template, forcefully);
        } else if (obj instanceof TemplateType type) {
            PocketCloud.instance().servers().stop(type, forcefully);
        } else if (obj instanceof ServerGroup group) {
            PocketCloud.instance().servers().stop(group, forcefully);
        } else if (obj instanceof ServerSoftware software) {
            for (ICloudServer server : PocketCloud.instance().servers().query(q -> q.runningSoftware(software.name()))) {
                PocketCloud.instance().servers().stop(server, forcefully);
            }
        }
        return true;
    }

    public boolean handleStopAllSub(CommandSender sender, CommandContext ctx) {
        boolean sure = ctx.hasFlag("y");
        boolean forcefully = ctx.hasFlag("f");
        if (sure) {
            PocketCloud.instance().servers().stopAll(forcefully);
        } else {
            awaitConfirmation(sender, "Are you sure you want to §cstop §rall servers?")
                    .thenSuccess(r -> {
                        if (r) {
                            PocketCloud.instance().servers().stopAll(forcefully);
                        }
                    });
        }
        return true;
    }

    public boolean handleDispatchSub(CommandSender sender, CommandContext ctx) {
        CloudServer server = ctx.arg("server", CloudServer.class);
        String command = ctx.arg("command", String.class);
        server.dispatch(command).thenSuccess(res -> {
            sender.success("Successfully ran the command on §b{}§r, server responded with the following messages:", server.name());
            if (res.messages().isEmpty()) sender.success("§cNone");
            for (String message : res.messages()) {
                for (String part : message.split("\n")) {
                    sender.success(part.trim());
                }
            }
        }).failure(t -> sender.error("Failed to dispatch command: §e{}", t.getMessage()));
        return true;
    }

    public boolean handleSaveSub(CommandSender sender, CommandContext ctx) {
        CloudServer server = ctx.arg("server", CloudServer.class);
        sender.info("Saving §b{}§r...", server.name());
        Instant start = Instant.now();
        server.save().thenSuccess(_ -> sender.success("Successfully §asaved §b{}§r. §8(§rTook §b{}ms§8)", server.name(), NumberUtils.formatNumber(Duration.between(start, Instant.now()).toMillis() / 1000F, 3)))
                .failure(e -> sender.error("Failed to save §b{}§r: §c{}", server.name(), e.getMessage()));
        return true;
    }

    public boolean handleInfoSub(CommandSender sender, CommandContext ctx) {
        CloudServer server = ctx.arg("server", CloudServer.class);
        sender.info("Name§8: §b{} §8(§b{}§8/§b{}§8)", server.name(), server.id(), server.templateName());
        sender.info("UUID§8: §b{}", server.uuid().toString());
        sender.info("Status§8: §b{} §8(§e{}§8)", server.status().getDisplay(), server.verificationStatus().name());
        sender.info("Port§8: §b{}", server.data().port());
        sender.info("Players§8: §b{}§8/§c{}", server.playerCount(), server.data().maxPlayers());
        sender.info("TPS§8: §b{} §8(§rAverage: §b{}§8)", FormatUtils.tps(server.data().tps(), true), FormatUtils.tps(server.data().avgTps(), true));
        sender.info("Memory Usage§8: §b{}§8/§c{} §8(§rPeak: §b{}§8)", FormatUtils.bytes(server.data().memoryUsage(), true), FormatUtils.bytes(server.data().memoryLimit(), false), FormatUtils.bytes(server.data().memoryPeak(), true));
        sender.info("CPU Usage§8: §b{}", FormatUtils.usagePercentage(server.data().cpuUsage(), true));
        sender.info("Used Software§8: §b{}", server.template().serverSoftware().name());
        sender.info("Path§8: §b{}", server.path().toAbsolutePath().toString());
        sender.info("Channel§8: §b{}", server.client().isPresent() ? server.client().get().address().toString() : "No channel yet");
        return true;
    }

    public boolean handleScreenSub(CommandSender sender, CommandContext ctx) {
        CloudServer server = ctx.arg("server", CloudServer.class);
        PocketCloud.instance().screens().set(new ServerConsoleMonitorScreen(server.name()));
        return true;
    }

    public boolean handleListSub(CommandSender sender, CommandContext ctx) {
        ServerSearchQuery query = ServerSearchQuery.create();
        Object filter = ctx.arg("filter");
        if (filter instanceof Template template) {
            query.ofTemplate(template);
        } else if (filter instanceof TemplateType type) {
            query.ofType(type);
        } else if (filter instanceof ServerGroup group) {
            query.inGroup(group);
        } else if (filter instanceof ServerSoftware software) {
            query.runningSoftware(software.name());
        }

        Collection<ICloudServer> servers = PocketCloud.instance().servers().query(query).stream().sorted(Comparator.comparingInt(ICloudServer::playerCount)).toList();
        sender.info("Servers §8(§b{}§8/§b{}§8)§r:", servers.size(), PocketCloud.instance().servers().getAll().size());
        if (servers.isEmpty()) sender.info("§cNo servers found.");
        for (ICloudServer server : servers) {
            sender.info("Name§8: §b{} §8| §rPlayers§8: §b{}§8/§c{} §8| §rPort§8: §b{} §8| §rStatus§8: §b{} §8| §rSoftware§8: §b{} §8(§b{}§8)", server.name(),
                    server.playerCount(), server.data().maxPlayers(), server.data().port(), server.status().getDisplay(),
                    server.template().serverSoftware().name(), server.template().templateType().name());
        }

        return true;
    }
}